Tagging ui

hardiesoft opened the pull request:
Bulk tagging UI prototype ready for initial review.  I've broken each individual commit down into more digestible pieces, please let me know if you have any questions.  Note that mobile layout hasn't had any love at this stage.

hardiesoft commented:
Hmm, looks like I need to figure out how to tell Jest how to transpile typescript, I guess it doesn't just use the project config for this.  Will look into this tonight.

mjs commented:
I haven't looked at the code yet but have the following feedback from using it:

- firstly, this is very cool!
- the looping over the current track is great
- it looks like boxes are being drawn for each track in the recording which makes it harder to determine what the current track is
- If you clikc "Skip track" when the Play again/Next Video controls are up, those controls don't go away and playback area remains greyed out
- There needs to be a way to access the full range of tags (i.e. the modal that appears when you use the + button on the current tagging UI
- A quick tag button for mustelids would be nice
- can motion paths just be shown for the current track?
- I wonder if Skip Track should be at the bottom of the quick tag buttons as skipping is just another final decision for the track
- Can the countdown for the next track be a little quicker?

mjs commented:
I've just been trying this out with prod recordings and it really is a significant improvement in terms of tagging efficiency!

hardiesoft commented:
- I've made boxes being drawn only for current track an optional feature of the overlay.  Note that it's still the default behaviour on the regular video player.
- Next/previous buttons at the end of the video shouldn't display when looping of the current track is enabled, fixed.
- For custom tags, I've added support for the current modal, but removed confidence and comment to make it a bit more succinct.  I'm not sure a modal is the best thing here though, or whether now that we're getting more tag buttons we should try and lay them out differently (we still need a different layout for this on mobile anyway)
- Motion tracks can now be made to only show for the current track, but I think generally it's a useful feature to be able to show them for all tracks, because it shows obviously where multiple tracks are really one track with breaks in it due to occlusion etc.
- Skip track is now just another action button grouping-wise, I agree this makes sense.
- Countdown is now quicker.
- Added ts support for jest.

I think generally once we are happy with the behaviour of the prototype, it would be good to make the icons for tags a bit more uniform in terms of size and style (or maybe we don't really care?)
Motion paths button probably wants to become part of the video player chrome, but that can be deferred to when I do more work on the video player as part of CPTV3 stuff if we want.

mjs commented on src/api/Recording.api.ts:
```diff
@@ -0,0 +1,243 @@
+}
+
+export interface TrackTag {
+  id: TrackTagId; // NOTE(jon): Different than a TagId?
```
Yes. A `TagId` would refer to a recording tag, not a track tag.

mjs commented on src/api/Recording.api.ts:
```diff
@@ -0,0 +1,243 @@
+  count: number;
+  limit: string; // NOTE(jon): Actually, a number, but comes back as a string...
+  messages: string[];
+  offset: string; // NOTE(jon): Actually, a number, but comes back as a string...
```
Well that sucks. I'll create a ticket for this.

mjs replied:
https://github.com/TheCacophonyProject/cacophony-api/issues/276

mjs commented on src/components/Video/VideoTracksScrubber.vue:
```diff
@@ -34,18 +35,27 @@
-<script>
+<script lang="ts">
+// eslint-disable-next-line no-unused-vars
+import { Track } from "../../api/Recording.api";
```
Does eslint not look in TS type signatures?

hardiesoft replied:
Turns out I needed to use the "@typescript-eslint/no-unused-vars": version of this lint in .eslintrc.js

mjs commented:
>     * I've made boxes being drawn only for current track an optional feature of the overlay.  Note that it's still the default behaviour on the regular video player.

:+1: 

>     * Next/previous buttons at the end of the video shouldn't display when looping of the current track is enabled, fixed.

:+1: 

>     * For custom tags, I've added support for the current modal, but removed confidence and comment to make it a bit more succinct.  I'm not sure a modal is the best thing here though, or whether now that we're getting more tag buttons we should try and lay them out differently (we still need a different layout for this on mobile anyway)

It would be nice to have more tags available as buttons but how it works now is fine for a start.

>     * Motion tracks can now be made to only show for the current track, but I think generally it's a useful feature to be able to show them for all tracks, because it shows obviously where multiple tracks are really one track with breaks in it due to occlusion etc.

I can see the benefit of showing all the tracks at once but it's really messy sometimes (depends on the recording). It would be good to have a way for the user to bring a single track or all of them.

>     * Countdown is now quicker.

Yep, that's better.

> I think generally once we are happy with the behaviour of the prototype, it would be good to make the icons for tags a bit more uniform in terms of size and style (or maybe we don't really care?)

It looks ok for now but we will probably want to get Sara involved at some point. We've already put a lot of other stuff on her plate though :)

> Motion paths button probably wants to become part of the video player chrome, but that can be deferred to when I do more work on the video player as part of CPTV3 stuff if we want.

Sounds good.

mjs commented:
I'm actually pretty happy for this to land soon. It's already quite useful.

Here's some remaining work as I see it (in rough order of priority):

- There's a bug if you hit "Skip track" in quick succession - the countdown starts showing negative numbers.
- Can we limit who can see the UI for now to just users with a globalPermission of "read" or "write"? I think that's already available in the frontend on the user object. That way the core team and some key people who already have elevated privs can kick the tires without everyone seeing it.
- Add a "skip recording" button which not only skips the recording but takes you to a completely different device and time (random jump). Sometimes you stuck in a fairly boring set of recordings and want a change. This can happen in a later PR.
- Switch to a dedicated API for bulk tagging instead of the using the existing APIs. This will fix the occasional delays when seeking to a new recording and allow anyone to tag recordings. This can happen in a later PR.
- A "delete recording" button which is only shown for users with a globalPermission of "write".  This can happen in a later PR.

mjs commented:
One more thing: why does it seem to be sequentially scanning to find a recording to tag? If you query for recordings that haven't been tagged by a human (this is a feature) you should get a focussed list of candidates.

mjs commented:
Let's also lose the "TM" on Power Tagger :)

mjs commented:
I see that when playing videos through the conventional recordings screen you now also get the Play Again?/Next Video buttons. This is ok but when you hit one of the arrow buttons underneath the played when these new overlaid buttons are displayed they don't disappear when the next video starts playing.

Example: https://photos.app.goo.gl/dyXBS9icddCqK1xe9

hardiesoft commented:
> 
> 
> One more thing: why does it seem to be sequentially scanning to find a recording to tag? If you query for recordings that haven't been tagged by a human (this is a feature) you should get a focussed list of candidates.

We do use that api, but remember I mentioned that we don't have an api to filter those recordings only to recordings that have *tracks*, so it's searching through the list until it finds one with tracks.  We decided not to treat recordings without tracks as being one big track I believe.

mjs commented:
> > One more thing: why does it seem to be sequentially scanning to find a recording to tag? If you query for recordings that haven't been tagged by a human (this is a feature) you should get a focussed list of candidates.
> 
> We do use that api, but remember I mentioned that we don't have an api to filter those recordings only to recordings that have _tracks_, so it's searching through the list until it finds one with tracks. We decided not to treat recordings without tracks as being one big track I believe.

Ah right. We must be hitting large chunks of track-less recordings sometimes.

hardiesoft commented:
I think I've addressed all the things that you said didn't need to be in a separate PR now, cheers.
